File: game/tools/editor_call.py
import urllib.request
import json
import sys

"""
Godot Editor Remote Executor — 标准调用脚本

用法：
    python editor_call.py <gdscript_code>
    python editor_call.py --file <path_to_script.gd>
    python editor_call.py --executors    # 仅列出已连接编辑器
    python editor_call.py --health       # 健康检查
    python editor_call.py --scene-tree   # 获取场景树

示例：
    python editor_call.py 'print("hello")'
    python editor_call.py --executors
"""

BASE_URL = "http://localhost:5302/api"
AUTH_TOKEN = "995e7c3f6fabc40a1bcd8a6f94dcad0106959c26c5827d2d3b261e1969109bd7"
HEADERS = {
    "Authorization": f"Bearer {AUTH_TOKEN}",
    "Content-Type": "application/json"
}

def get_executors():
    """获取已连接编辑器列表"""
    req = urllib.request.Request(f"{BASE_URL}/executors", headers=HEADERS)
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read().decode("utf-8"))

def get_first_executor_id():
    """获取第一个已连接编辑器的 executor_id"""
    data = get_executors()
    if data.get("success") and data.get("data"):
        return data["data"][0]["id"]
    return None

def execute(code, executor_id=None):
    """在 Godot 编辑器中执行 GDScript 代码"""
    if executor_id is None:
        executor_id = get_first_executor_id()
        if executor_id is None:
            print("[ERROR] No Godot editor connected. Start broker-server and enable plugin.", file=sys.stderr)
            return None

    body = json.dumps({"code": code, "executor_id": executor_id}).encode("utf-8")
    req = urllib.request.Request(
        f"{BASE_URL}/execute",
        data=body,
        headers=HEADERS,
        method="POST"
    )
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read().decode("utf-8"))

def format_output(outputs):
    """格式化 executeContext.output 的输出"""
    for out in outputs:
        key = out[0] if isinstance(out, list) else out
        val = out[1] if isinstance(out, list) else outputs[out]
        print(f"  {key} = {val}")

def main():
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)

    if sys.argv[1] == "--executors":
        data = get_executors()
        if data.get("success"):
            for ed in data.get("data", []):
                print(f"  ID: {ed['id']}")
                print(f"  Project: {ed.get('project_name', 'N/A')}")
                print(f"  PID: {ed.get('editor_pid', 'N/A')}")
                print(f"  Status: {ed.get('status', 'N/A')}")
                print()
        else:
            print(f"[ERROR] {data}")
        return

    if sys.argv[1] == "--health":
        req = urllib.request.Request(f"{BASE_URL}/health")
        with urllib.request.urlopen(req) as resp:
            print(resp.read().decode("utf-8"))
        return

    if sys.argv[1] == "--scene-tree":
        req = urllib.request.Request(f"{BASE_URL}/scene/tree", headers=HEADERS)
        with urllib.request.urlopen(req) as resp:
            data = json.loads(resp.read().decode("utf-8"))
            print(json.dumps(data, indent=2, ensure_ascii=False))
        return

    if sys.argv[1] == "--file" and len(sys.argv) >= 3:
        with open(sys.argv[2], encoding="utf-8") as f:
            code = f.read()
    else:
        code = " ".join(sys.argv[1:])

    print(f"[Executing GDScript...]")
    result = execute(code)

    if result is None:
        return

    data = result.get("data", {})

    if not data.get("compile_success"):
        print(f"[COMPILE ERROR] {data.get('compile_error', 'Unknown')}", file=sys.stderr)
        sys.exit(1)

    if not data.get("run_success"):
        print(f"[RUN ERROR] {data.get('run_error', 'Unknown')}", file=sys.stderr)
        sys.exit(1)

    print("[SUCCESS]")
    outputs = data.get("outputs", [])
    if outputs:
        format_output(outputs)

if __name__ == "__main__":
    main()

File: game/tools/editor_call.js
#!/usr/bin/env node
/**
 * editor_call.js — Godot Editor Remote Executor (Node.js 版 · 备选)
 *
 * 主推方案：python tools/editor_call.py（Python 3.12）
 * 本文件仅在 Python 不可用时作为 fallback。
 *
 * 与 Python 版功能完全一致：
 * - 自动获取 executor_id
 * - 用 \t 缩进编写 GDScript（避免空格陷阱）
 * - 正确处理 executeContext.output 数组格式
 * - 清晰的错误处理
 *
 * 用法:
 *   node editor_call.js 'print("hello")'          # 直接执行代码
 *   node editor_call.js --file script.gd           # 从文件执行
 *   node editor_call.js --executors                # 列出已连接编辑器
 *   node editor_call.js --health                   # 健康检查
 *   node editor_call.js --scene-tree               # 获取场景树
 */

const BASE_URL = "http://localhost:5302/api";
const AUTH_TOKEN = "995e7c3f6fabc40a1bcd8a6f94dcad0106959c26c5827d2d3b261e1969109bd7";

const http = require("http");
const fs = require("fs");
const path = require("path");

function apiCall(endpoint, method = "GET", body = null) {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: "localhost",
      port: 5302,
      path: `/api${endpoint}`,
      method,
      headers: {
        Authorization: `Bearer ${AUTH_TOKEN}`,
      },
    };
    if (body) {
      options.headers["Content-Type"] = "application/json";
    }
    const req = http.request(options, (res) => {
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => {
        try {
          resolve(JSON.parse(data));
        } catch {
          reject(new Error(`Invalid JSON: ${data}`));
        }
      });
    });
    req.on("error", reject);
    if (body) req.write(JSON.stringify(body));
    req.end();
  });
}

async function getFirstExecutorId() {
  const data = await apiCall("/executors");
  if (data.success && Array.isArray(data.data) && data.data.length > 0) {
    return data.data[0].id;
  }
  return null;
}

async function execute(code, executorId = null) {
  if (!executorId) {
    executorId = await getFirstExecutorId();
    if (!executorId) {
      console.error("[ERROR] No Godot editor connected. Start broker-server and enable plugin.");
      process.exit(1);
    }
  }
  return apiCall("/execute", "POST", { code, executor_id: executorId, timeout_ms: 30000 });
}

function formatOutput(outputs) {
  for (const out of outputs) {
    const key = out[0];
    const val = out[1];
    console.log(`  ${key} = ${val}`);
  }
}

async function main() {
  const args = process.argv.slice(2);
  if (args.length === 0) {
    showHelp();
    return;
  }

  const cmd = args[0];

  switch (cmd) {
    case "--executors": {
      const data = await apiCall("/executors");
      if (data.success) {
        for (const ed of data.data || []) {
          console.log(`  ID: ${ed.id}`);
          console.log(`  Project: ${ed.project_name || "N/A"}`);
          console.log(`  PID: ${ed.editor_pid || "N/A"}`);
          console.log(`  Status: ${ed.status || "N/A"}`);
          console.log();
        }
      } else {
        console.error(`[ERROR] ${JSON.stringify(data)}`);
      }
      break;
    }

    case "--health": {
      const data = await apiCall("/health");
      console.log(JSON.stringify(data, null, 2));
      break;
    }

    case "--scene-tree": {
      const data = await apiCall("/scene/tree");
      console.log(JSON.stringify(data, null, 2));
      break;
    }

    case "--file": {
      if (args.length < 2) {
        console.error("[ERROR] Usage: node editor_call.js --file <path>");
        process.exit(1);
      }
      const code = fs.readFileSync(args[1], "utf8");
      const result = await execute(code);
      printResult(result);
      break;
    }

    default: {
      const code = args.join(" ");
      const result = await execute(code);
      printResult(result);
    }
  }
}

function printResult(result) {
  if (!result) return;
  const data = result.data || {};

  if (!data.compile_success) {
    console.error(`[COMPILE ERROR] ${data.compile_error || "Unknown"}`);
    process.exit(1);
  }

  if (!data.run_success) {
    console.error(`[RUN ERROR] ${data.run_error || "Unknown"}`);
    process.exit(1);
  }

  console.log("[SUCCESS]");
  const outputs = data.outputs || [];
  if (outputs.length > 0) {
    formatOutput(outputs);
  }
}

function showHelp() {
  console.log(`
Godot Editor Remote Executor — 标准调用脚本

用法:
    node editor_call.js <gdscript_code>
    node editor_call.js --file <path_to_script.gd>
    node editor_call.js --executors    # 仅列出已连接编辑器
    node editor_call.js --health       # 健康检查
    node editor_call.js --scene-tree   # 获取场景树

示例:
    node editor_call.js 'print("hello")'
    node editor_call.js --executors

注意: GDScript 缩进必须用 Tab(\\t)，不能用空格
  `);
}

main().catch((err) => {
  console.error(`[ERROR] ${err.message}`);
  process.exit(1);
});
